In [141]:
# Importing the libraries

import pandas as pd
import numpy as np
import warnings
import openpyxl
import datetime 
import sys
sys.path.append("..")
warnings.filterwarnings('ignore')
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import FunctionTransformer

In [142]:
# Reading the call center dataset

call_center = pd.read_excel(r'C:\Users\hp\OneDrive - students.jkuat.ac.ke\Desktop\Tiko\data_call_center.xlsx')
call_center.head()

start                     end  Rafiki ID  \
0 2024-06-29 18:01:28.770 2024-06-29 18:01:43.200    1920703   
1 2024-06-29 18:01:44.724 2024-06-29 18:03:28.992    2016625   
2 2024-06-29 18:03:31.055 2024-06-29 18:03:40.289    2013477   
3 2024-06-29 18:03:42.239 2024-06-29 18:03:53.183    2016141   
4 2024-06-29 18:03:55.123 2024-06-29 18:04:12.694    1970498   

  Month of Service Uptake    Call status What is your gender?  \
0              2024-06-01  Not reachable                  NaN   
1              2024-06-01  Not reachable                  NaN   
2              2024-06-01  Not reachable                  NaN   
3              2024-06-01  Not reachable                  NaN   
4              2024-06-01  Not reachable                  NaN   

  What year were you born?  \
0                      NaN   
1                      NaN   
2                      NaN   
3                      NaN   
4                      NaN   

  Did you recently visit a Tiko health facility to access a service?  \
0                                                NaN                   
1                                                NaN                   
2                                                NaN                   
3                                                NaN                   
4                                                NaN                   

  Were you given a tiko card? Were you satisfied with the service you took?  \
0                         NaN                                           NaN   
1                         NaN                                           NaN   
2                         NaN                                           NaN   
3                         NaN                                           NaN   
4                         NaN                                           NaN   

  Please share any other feedback you have concerning the service you took, or the facility (clinic/chemist) you visited  \
0                                                NaN                                                                       
1                                                NaN                                                                       
2                                                NaN                                                                       
3                                                NaN                                                                       
4                                                NaN                                                                       

  Do you have any additional questions or concerns about family planning or Tiko?  \
0                                                NaN                                
1                                                NaN                                
2                                                NaN                                
3                                                NaN                                
4                                                NaN                                

  Write any intelligence collected from the call  
0                                            NaN  
1                                            NaN  
2                                            NaN  
3                                            NaN  
4                                            NaN

In [143]:
call_center.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1389 entries, 0 to 1388
Data columns (total 13 columns):
 #   Column                                                                                                                  Non-Null Count  Dtype         
---  ------                                                                                                                  --------------  -----         
 0   start                                                                                                                   1389 non-null   datetime64[ns]
 1   end                                                                                                                     1389 non-null   datetime64[ns]
 2   Rafiki ID                                                                                                               1389 non-null   int64         
 3   Month of Service Uptake                                                                                     

In [144]:
# Reading the tiko platform dataset

tiko_platform = pd.read_excel(r'C:\Users\hp\OneDrive - students.jkuat.ac.ke\Desktop\Tiko\data_tiko_platform.xlsx')
tiko_platform.head()

Offer Redemption Event Id  Rafiki Id                 Offer Trader Name  \
0                  487759949    1329043       Mathare North Health Centre   
1                  487761498    2015957   Medwell Pharmaceuticals Masimba   
2                  487826898    2014933                   Medlyf Pharmacy   
3                  487843287    2009023  Lubryl Medical & Consultancy Ltd   
4                  487911292    2016970          Lungalunga Health Centre   

  Enrolment Path Classifier            Enrolment Type  \
0            SMS - Low Tech  Enrolment With Mobiliser   
1            SMS - Low Tech  Enrolment With Mobiliser   
2            SMS - Low Tech  Enrolment With Mobiliser   
3            SMS - Low Tech  Enrolment With Mobiliser   
4            SMS - Low Tech  Enrolment With Mobiliser   

  Offer Trader County Name Offer Trader Business Organisation  \
0                  Nairobi           Public Sector Facilities   
1                  Nairobi                           Pharmacy   
2                   Kisumu                           Pharmacy   
3                  Nairobi     Private Non Franchised Clinics   
4                  Nairobi           Public Sector Facilities   

  Offer Redemption Date Service Type  
0   26/06/2024 08:30:14          HIV  
1   26/06/2024 08:33:33          HST  
2   26/06/2024 10:00:56          HST  
3   26/06/2024 10:15:20           OT  
4   26/06/2024 11:01:13           OC

In [145]:
tiko_platform.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1512 entries, 0 to 1511
Data columns (total 9 columns):
 #   Column                              Non-Null Count  Dtype 
---  ------                              --------------  ----- 
 0   Offer Redemption Event Id           1512 non-null   int64 
 1   Rafiki Id                           1512 non-null   int64 
 2   Offer Trader Name                   1512 non-null   object
 3   Enrolment Path Classifier           1512 non-null   object
 4   Enrolment Type                      1512 non-null   object
 5   Offer Trader County Name            1512 non-null   object
 6   Offer Trader Business Organisation  1512 non-null   object
 7   Offer Redemption Date               1512 non-null   object
 8   Service Type                        1512 non-null   object
dtypes: int64(2), object(7)
memory usage: 106.4+ KB


In [146]:
tiko_platform.rename(columns={
                   'Offer Redemption Event Id': 'Event Id'}, inplace=True)
tiko_platform.head()

Event Id  Rafiki Id                 Offer Trader Name  \
0  487759949    1329043       Mathare North Health Centre   
1  487761498    2015957   Medwell Pharmaceuticals Masimba   
2  487826898    2014933                   Medlyf Pharmacy   
3  487843287    2009023  Lubryl Medical & Consultancy Ltd   
4  487911292    2016970          Lungalunga Health Centre   

  Enrolment Path Classifier            Enrolment Type  \
0            SMS - Low Tech  Enrolment With Mobiliser   
1            SMS - Low Tech  Enrolment With Mobiliser   
2            SMS - Low Tech  Enrolment With Mobiliser   
3            SMS - Low Tech  Enrolment With Mobiliser   
4            SMS - Low Tech  Enrolment With Mobiliser   

  Offer Trader County Name Offer Trader Business Organisation  \
0                  Nairobi           Public Sector Facilities   
1                  Nairobi                           Pharmacy   
2                   Kisumu                           Pharmacy   
3                  Nairobi     Private Non Franchised Clinics   
4                  Nairobi           Public Sector Facilities   

  Offer Redemption Date Service Type  
0   26/06/2024 08:30:14          HIV  
1   26/06/2024 08:33:33          HST  
2   26/06/2024 10:00:56          HST  
3   26/06/2024 10:15:20           OT  
4   26/06/2024 11:01:13           OC

In [147]:
financial_impact = pd.read_excel(r'C:\Users\hp\OneDrive - students.jkuat.ac.ke\Desktop\Tiko\data_financial_impact.xlsx')
financial_impact.head()

offer_redemption_event_id Actor Role  Tiko Amount
0                  487759949      Agent          125
1                  487759949   Provider           20
2                  487759949     Rafiki           60
3                  487761498   Provider          288
4                  487826898   Provider          288

In [148]:
financial_impact.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4055 entries, 0 to 4054
Data columns (total 3 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   offer_redemption_event_id  4055 non-null   int64 
 1   Actor Role                 4055 non-null   object
 2   Tiko Amount                4055 non-null   int64 
dtypes: int64(2), object(1)
memory usage: 95.2+ KB


In [149]:
# Renaming offer id field

financial_impact.rename(columns={
                   'offer_redemption_event_id': 'Event Id'}, inplace=True)
financial_impact.head()

Event Id Actor Role  Tiko Amount
0  487759949      Agent          125
1  487759949   Provider           20
2  487759949     Rafiki           60
3  487761498   Provider          288
4  487826898   Provider          288

In [150]:
# Finding the missing values in call center dataset

call_center.isna().sum()

start                                                                                                                        0
end                                                                                                                          0
Rafiki ID                                                                                                                    0
Month of Service Uptake                                                                                                      0
Call status                                                                                                                  0
What is your gender?                                                                                                      1248
What year were you born?                                                                                                  1249
Did you recently visit a Tiko health facility to access a service?                                             

In [151]:
# Fixing the missing values with Unknown

call_center['What is your gender?'] = call_center['What is your gender?'].fillna('Unknown')
call_center['What year were you born?'] = call_center['What year were you born?'].fillna('Unknown')
call_center['Did you recently visit a Tiko health facility to access a service?'] = call_center['Did you recently visit a Tiko health facility to access a service?'].fillna('Unknown')
call_center['Were you given a tiko card?'] = call_center['Were you given a tiko card?'].fillna('Unknown')
call_center['Were you satisfied with the service you took?'] = call_center['Were you satisfied with the service you took?'].fillna('Not provided')
call_center['Please share any other feedback you have concerning the service you took, or the facility (clinic/chemist) you visited'] = call_center['Please share any other feedback you have concerning the service you took, or the facility (clinic/chemist) you visited'].fillna('Not provided')
call_center['Do you have any additional questions or concerns about family planning or Tiko?'] = call_center['Do you have any additional questions or concerns about family planning or Tiko?'].fillna('Not provided')
call_center['Write any intelligence collected from the call'] = call_center['Write any intelligence collected from the call'].fillna('Not provided')


In [152]:
# Confirming missing values in call center data

call_center.isna().sum()

start                                                                                                                     0
end                                                                                                                       0
Rafiki ID                                                                                                                 0
Month of Service Uptake                                                                                                   0
Call status                                                                                                               0
What is your gender?                                                                                                      0
What year were you born?                                                                                                  0
Did you recently visit a Tiko health facility to access a service?                                                        0
Were you

In [153]:
# Finding missing values in tiko platform data

tiko_platform.isna().sum()

Event Id                              0
Rafiki Id                             0
Offer Trader Name                     0
Enrolment Path Classifier             0
Enrolment Type                        0
Offer Trader County Name              0
Offer Trader Business Organisation    0
Offer Redemption Date                 0
Service Type                          0
dtype: int64

In [154]:
# Finding missing values in financial_impact data

financial_impact.isna().sum()

Event Id       0
Actor Role     0
Tiko Amount    0
dtype: int64

In [155]:
# Renaming column names in call center data

call_center.rename(columns={'Rafiki ID': 'Rafiki Id',
                   'What is your gender?': 'Gender', 
                   'What year were you born?': 'Year of birth',
                   'Did you recently visit a Tiko health facility to access a service?': 'Visited facility',
                   'Were you given a tiko card?': 'Tiko card',
                   'Were you satisfied with the service you took?': 'Satisfaction',
                   'Please share any other feedback you have concerning the service you took, or the facility (clinic/chemist) you visited': 'Feedback',
                   'Do you have any additional questions or concerns about family planning or Tiko?': 'Concerns',
                   'Write any intelligence collected from the call': 'Notes'}, inplace=True)
call_center.head()

start                     end  Rafiki Id  \
0 2024-06-29 18:01:28.770 2024-06-29 18:01:43.200    1920703   
1 2024-06-29 18:01:44.724 2024-06-29 18:03:28.992    2016625   
2 2024-06-29 18:03:31.055 2024-06-29 18:03:40.289    2013477   
3 2024-06-29 18:03:42.239 2024-06-29 18:03:53.183    2016141   
4 2024-06-29 18:03:55.123 2024-06-29 18:04:12.694    1970498   

  Month of Service Uptake    Call status   Gender Year of birth  \
0              2024-06-01  Not reachable  Unknown       Unknown   
1              2024-06-01  Not reachable  Unknown       Unknown   
2              2024-06-01  Not reachable  Unknown       Unknown   
3              2024-06-01  Not reachable  Unknown       Unknown   
4              2024-06-01  Not reachable  Unknown       Unknown   

  Visited facility Tiko card  Satisfaction      Feedback      Concerns  \
0          Unknown   Unknown  Not provided  Not provided  Not provided   
1          Unknown   Unknown  Not provided  Not provided  Not provided   
2          Unknown   Unknown  Not provided  Not provided  Not provided   
3          Unknown   Unknown  Not provided  Not provided  Not provided   
4          Unknown   Unknown  Not provided  Not provided  Not provided   

          Notes  
0  Not provided  
1  Not provided  
2  Not provided  
3  Not provided  
4  Not provided

In [156]:
# Dropping irrelevant columns

call_center = call_center.drop(['start', 'end', 'Notes'], axis=1)
call_center = call_center.loc[:, call_center.apply(pd.Series.nunique) != 1]
call_center.head()

Rafiki Id Month of Service Uptake    Call status   Gender Year of birth  \
0    1920703              2024-06-01  Not reachable  Unknown       Unknown   
1    2016625              2024-06-01  Not reachable  Unknown       Unknown   
2    2013477              2024-06-01  Not reachable  Unknown       Unknown   
3    2016141              2024-06-01  Not reachable  Unknown       Unknown   
4    1970498              2024-06-01  Not reachable  Unknown       Unknown   

  Visited facility Tiko card  Satisfaction      Feedback      Concerns  
0          Unknown   Unknown  Not provided  Not provided  Not provided  
1          Unknown   Unknown  Not provided  Not provided  Not provided  
2          Unknown   Unknown  Not provided  Not provided  Not provided  
3          Unknown   Unknown  Not provided  Not provided  Not provided  
4          Unknown   Unknown  Not provided  Not provided  Not provided

In [157]:
call_center['Month of Service Uptake'].unique()


<DatetimeArray>
['2024-06-01 00:00:00', '2024-02-01 00:00:00', '2024-07-01 00:00:00',
 '2024-05-01 00:00:00']
Length: 4, dtype: datetime64[ns]

In [158]:
# Extract the month from the datetime column 

call_center['Month'] = call_center['Month of Service Uptake'].dt.month_name()
call_center.head()


Rafiki Id Month of Service Uptake    Call status   Gender Year of birth  \
0    1920703              2024-06-01  Not reachable  Unknown       Unknown   
1    2016625              2024-06-01  Not reachable  Unknown       Unknown   
2    2013477              2024-06-01  Not reachable  Unknown       Unknown   
3    2016141              2024-06-01  Not reachable  Unknown       Unknown   
4    1970498              2024-06-01  Not reachable  Unknown       Unknown   

  Visited facility Tiko card  Satisfaction      Feedback      Concerns Month  
0          Unknown   Unknown  Not provided  Not provided  Not provided  June  
1          Unknown   Unknown  Not provided  Not provided  Not provided  June  
2          Unknown   Unknown  Not provided  Not provided  Not provided  June  
3          Unknown   Unknown  Not provided  Not provided  Not provided  June  
4          Unknown   Unknown  Not provided  Not provided  Not provided  June

In [159]:
# Confirming the months

call_center['Month'].unique()

array(['June', 'February', 'July', 'May'], dtype=object)

In [160]:
# Deleting duplicate field

del call_center['Month of Service Uptake']
call_center.head()

Rafiki Id    Call status   Gender Year of birth Visited facility Tiko card  \
0    1920703  Not reachable  Unknown       Unknown          Unknown   Unknown   
1    2016625  Not reachable  Unknown       Unknown          Unknown   Unknown   
2    2013477  Not reachable  Unknown       Unknown          Unknown   Unknown   
3    2016141  Not reachable  Unknown       Unknown          Unknown   Unknown   
4    1970498  Not reachable  Unknown       Unknown          Unknown   Unknown   

   Satisfaction      Feedback      Concerns Month  
0  Not provided  Not provided  Not provided  June  
1  Not provided  Not provided  Not provided  June  
2  Not provided  Not provided  Not provided  June  
3  Not provided  Not provided  Not provided  June  
4  Not provided  Not provided  Not provided  June

In [161]:
# Creating a clean call center data file

call_center.to_excel(r'C:\Users\hp\OneDrive - students.jkuat.ac.ke\Desktop\Tiko\call_center_clean.xlsx',index=False)

# Naming the workbook sheet

ss = openpyxl.load_workbook(r"C:\Users\hp\OneDrive - students.jkuat.ac.ke\Desktop\Tiko\call_center_clean.xlsx")
ss_sheet = ss['Sheet1']
ss_sheet.title = 'call_center_clean'
ss.save(r"C:\Users\hp\OneDrive - students.jkuat.ac.ke\Desktop\Tiko\call_center_clean.xlsx")


In [162]:
# Creating a clean tiko platform data file

tiko_platform.to_excel(r'C:\Users\hp\OneDrive - students.jkuat.ac.ke\Desktop\Tiko\tiko_platform_clean.xlsx',index=False)

# Naming the workbook sheet

ss = openpyxl.load_workbook(r"C:\Users\hp\OneDrive - students.jkuat.ac.ke\Desktop\Tiko\tiko_platform_clean.xlsx")
ss_sheet = ss['Sheet1']
ss_sheet.title = 'tiko_platform_clean'
ss.save(r"C:\Users\hp\OneDrive - students.jkuat.ac.ke\Desktop\Tiko\tiko_platform_clean.xlsx")


In [163]:
# Creating a clean financial impcat data file

financial_impact.to_excel(r'C:\Users\hp\OneDrive - students.jkuat.ac.ke\Desktop\Tiko\financial_impact_clean.xlsx',index=False)

# Naming the workbook sheet

ss = openpyxl.load_workbook(r"C:\Users\hp\OneDrive - students.jkuat.ac.ke\Desktop\Tiko\financial_impact_clean.xlsx")
ss_sheet = ss['Sheet1']
ss_sheet.title = 'financial_impact_clean'
ss.save(r"C:\Users\hp\OneDrive - students.jkuat.ac.ke\Desktop\Tiko\financial_impact_clean.xlsx")


In [164]:
# Merging the datasets

call_center_clean = pd.read_excel(r'C:\Users\hp\OneDrive - students.jkuat.ac.ke\Desktop\Tiko\call_center_clean.xlsx')
tiko_platform_clean = pd.read_excel(r'C:\Users\hp\OneDrive - students.jkuat.ac.ke\Desktop\Tiko\tiko_platform_clean.xlsx')

df1 = pd.DataFrame(call_center_clean)
df2 = pd.DataFrame(tiko_platform_clean)

call_platform = pd.merge(df1, df2, on = "Rafiki Id")

# Creating a new file for the merged data

call_platform.to_excel(r'C:\Users\hp\OneDrive - students.jkuat.ac.ke\Desktop\Tiko\call_platform.xlsx',index=False)

# Reading the dataset
call_platform.head()


Rafiki Id    Call status   Gender Year of birth Visited facility Tiko card  \
0    1920703  Not reachable  Unknown       Unknown          Unknown   Unknown   
1    2016625  Not reachable  Unknown       Unknown          Unknown   Unknown   
2    2013477  Not reachable  Unknown       Unknown          Unknown   Unknown   
3    2016141  Not reachable  Unknown       Unknown          Unknown   Unknown   
4    1970498  Not reachable  Unknown       Unknown          Unknown   Unknown   

   Satisfaction      Feedback      Concerns Month   Event Id  \
0  Not provided  Not provided  Not provided  June  488030850   
1  Not provided  Not provided  Not provided  June  488033730   
2  Not provided  Not provided  Not provided  June  488035259   
3  Not provided  Not provided  Not provided  June  488035835   
4  Not provided  Not provided  Not provided  June  488038525   

               Offer Trader Name Enrolment Path Classifier  \
0             Nyalenda Gods Will            SMS - Low Tech   
1         Doruex Medicare Centre            SMS - Low Tech   
2          Umoja 1 Health Centre            SMS - Low Tech   
3  Menich Family Health Services            SMS - Low Tech   
4                Tassia Hospital            SMS - Low Tech   

             Enrolment Type Offer Trader County Name  \
0  Enrolment With Mobiliser                   Kisumu   
1  Enrolment With Mobiliser                    Busia   
2  Enrolment With Mobiliser                  Nairobi   
3       Enrolment With Peer                  Nairobi   
4  Enrolment With Mobiliser                  Nairobi   

  Offer Trader Business Organisation Offer Redemption Date Service Type  
0                               Amua   26/06/2024 12:14:28           OC  
1                                Rhn   26/06/2024 12:15:45          IMP  
2           Public Sector Facilities   26/06/2024 12:16:17           OC  
3     Private Non Franchised Clinics   26/06/2024 12:16:43          IMP  
4     Private Non Franchised Clinics   26/06/2024 12:18:10          HIV

In [165]:
# Merging call platform data with the financial data

call_platform = pd.read_excel(r'C:\Users\hp\OneDrive - students.jkuat.ac.ke\Desktop\Tiko\call_platform.xlsx')
financial_impact_clean = pd.read_excel(r'C:\Users\hp\OneDrive - students.jkuat.ac.ke\Desktop\Tiko\financial_impact_clean.xlsx')

df1 = pd.DataFrame(call_platform)
df2 = pd.DataFrame(financial_impact_clean)

merged_data = pd.merge(df1, df2, on = "Event Id")

# Creating a new file

merged_data.to_excel(r'C:\Users\hp\OneDrive - students.jkuat.ac.ke\Desktop\Tiko\merged_data.xlsx',index=False)

# Reading the merged data

merged_data.head()


Rafiki Id    Call status   Gender Year of birth Visited facility Tiko card  \
0    1920703  Not reachable  Unknown       Unknown          Unknown   Unknown   
1    1920703  Not reachable  Unknown       Unknown          Unknown   Unknown   
2    1920703  Not reachable  Unknown       Unknown          Unknown   Unknown   
3    2016625  Not reachable  Unknown       Unknown          Unknown   Unknown   
4    2016625  Not reachable  Unknown       Unknown          Unknown   Unknown   

   Satisfaction      Feedback      Concerns Month   Event Id  \
0  Not provided  Not provided  Not provided  June  488030850   
1  Not provided  Not provided  Not provided  June  488030850   
2  Not provided  Not provided  Not provided  June  488030850   
3  Not provided  Not provided  Not provided  June  488033730   
4  Not provided  Not provided  Not provided  June  488033730   

        Offer Trader Name Enrolment Path Classifier            Enrolment Type  \
0      Nyalenda Gods Will            SMS - Low Tech  Enrolment With Mobiliser   
1      Nyalenda Gods Will            SMS - Low Tech  Enrolment With Mobiliser   
2      Nyalenda Gods Will            SMS - Low Tech  Enrolment With Mobiliser   
3  Doruex Medicare Centre            SMS - Low Tech  Enrolment With Mobiliser   
4  Doruex Medicare Centre            SMS - Low Tech  Enrolment With Mobiliser   

  Offer Trader County Name Offer Trader Business Organisation  \
0                   Kisumu                               Amua   
1                   Kisumu                               Amua   
2                   Kisumu                               Amua   
3                    Busia                                Rhn   
4                    Busia                                Rhn   

  Offer Redemption Date Service Type Actor Role  Tiko Amount  
0   26/06/2024 12:14:28           OC      Agent          125  
1   26/06/2024 12:14:28           OC   Provider          150  
2   26/06/2024 12:14:28           OC     Rafiki           60  
3   26/06/2024 12:15:45          IMP      Agent           50  
4   26/06/2024 12:15:45          IMP   Provider          690

In [166]:
# Investigating the merged dataset

merged_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4134 entries, 0 to 4133
Data columns (total 20 columns):
 #   Column                              Non-Null Count  Dtype 
---  ------                              --------------  ----- 
 0   Rafiki Id                           4134 non-null   int64 
 1   Call status                         4134 non-null   object
 2   Gender                              4134 non-null   object
 3   Year of birth                       4134 non-null   object
 4   Visited facility                    4134 non-null   object
 5   Tiko card                           4134 non-null   object
 6   Satisfaction                        4134 non-null   object
 7   Feedback                            4134 non-null   object
 8   Concerns                            4134 non-null   object
 9   Month                               4134 non-null   object
 10  Event Id                            4134 non-null   int64 
 11  Offer Trader Name                   4134 non-null   obje

In [167]:
merged_data.isnull().sum()


Rafiki Id                             0
Call status                           0
Gender                                0
Year of birth                         0
Visited facility                      0
Tiko card                             0
Satisfaction                          0
Feedback                              0
Concerns                              0
Month                                 0
Event Id                              0
Offer Trader Name                     0
Enrolment Path Classifier             0
Enrolment Type                        0
Offer Trader County Name              0
Offer Trader Business Organisation    0
Offer Redemption Date                 0
Service Type                          0
Actor Role                            0
Tiko Amount                           0
dtype: int64

In [168]:
# Calculate the percentage of missing values in each column
missing_percentages = (merged_data.isnull().sum() / len(merged_data)) * 100
print("\
Percentage of missing values in each column:")
print(missing_percentages[missing_percentages > 0])

Percentage of missing values in each column:
Series([], dtype: float64)


Based on this information, the dataset has 4,134 entries and 21 columns. Some columns have missing values ().

I'll proceed to clean it by;
1. Handling missing values
2. Converting data types where necessary
3. Removing any duplicate entries and irrelevant columns
4. Standardizing text data (e.g., capitalize consistently)

In [169]:
# Checking for unique values in year of birth column
merged_data['Year of birth'].unique()

array(['Unknown', '2004-2009', '1998-2003', 'Before 1998'], dtype=object)

But first, as you see above, it seems that the "Year of birth" column contains non-numeric values in ranges such as '2004-2009'. I will clean this column by converting valid year ranges to a single year or removing invalid entries.

In [170]:
# Replace year ranges with approximate median values
merged_data['Year of birth'].replace({'2004-2009': 2006, '1998-2003': 2000, 'Before 1998': 1997}, inplace=True)

# Convert 'Year of birth' to numeric, forcing errors to NaN
merged_data['Year of birth'] = pd.to_numeric(merged_data['Year of birth'], errors='coerce')

# Fill NaN values with the median of the cleaned numeric values
merged_data['Year of birth'].fillna(merged_data['Year of birth'].median(), inplace=True)


I'll proceed with cleaning the data

In [171]:
# Replace missing values for other columns
merged_data['Gender'].fillna('Unknown', inplace=True)
merged_data['Visited facility'].fillna('Unknown', inplace=True)
merged_data['Tiko card'].fillna('Unknown', inplace=True)
merged_data['Satisfaction'].fillna('Not provided', inplace=True)
merged_data['Feedback'].fillna('Not provided', inplace=True)
merged_data['Concerns'].fillna('None reported', inplace=True)


In [172]:
# Confirming missing values

merged_data.isnull().sum()

Rafiki Id                             0
Call status                           0
Gender                                0
Year of birth                         0
Visited facility                      0
Tiko card                             0
Satisfaction                          0
Feedback                              0
Concerns                              0
Month                                 0
Event Id                              0
Offer Trader Name                     0
Enrolment Path Classifier             0
Enrolment Type                        0
Offer Trader County Name              0
Offer Trader Business Organisation    0
Offer Redemption Date                 0
Service Type                          0
Actor Role                            0
Tiko Amount                           0
dtype: int64

In [173]:
# Deleting the rafiki Id and Event Id columns
del merged_data['Rafiki Id'],
del merged_data['Event Id']
merged_data.head()

Call status   Gender  Year of birth Visited facility Tiko card  \
0  Not reachable  Unknown         2006.0          Unknown   Unknown   
1  Not reachable  Unknown         2006.0          Unknown   Unknown   
2  Not reachable  Unknown         2006.0          Unknown   Unknown   
3  Not reachable  Unknown         2006.0          Unknown   Unknown   
4  Not reachable  Unknown         2006.0          Unknown   Unknown   

   Satisfaction      Feedback      Concerns Month       Offer Trader Name  \
0  Not provided  Not provided  Not provided  June      Nyalenda Gods Will   
1  Not provided  Not provided  Not provided  June      Nyalenda Gods Will   
2  Not provided  Not provided  Not provided  June      Nyalenda Gods Will   
3  Not provided  Not provided  Not provided  June  Doruex Medicare Centre   
4  Not provided  Not provided  Not provided  June  Doruex Medicare Centre   

  Enrolment Path Classifier            Enrolment Type  \
0            SMS - Low Tech  Enrolment With Mobiliser   
1            SMS - Low Tech  Enrolment With Mobiliser   
2            SMS - Low Tech  Enrolment With Mobiliser   
3            SMS - Low Tech  Enrolment With Mobiliser   
4            SMS - Low Tech  Enrolment With Mobiliser   

  Offer Trader County Name Offer Trader Business Organisation  \
0                   Kisumu                               Amua   
1                   Kisumu                               Amua   
2                   Kisumu                               Amua   
3                    Busia                                Rhn   
4                    Busia                                Rhn   

  Offer Redemption Date Service Type Actor Role  Tiko Amount  
0   26/06/2024 12:14:28           OC      Agent          125  
1   26/06/2024 12:14:28           OC   Provider          150  
2   26/06/2024 12:14:28           OC     Rafiki           60  
3   26/06/2024 12:15:45          IMP      Agent           50  
4   26/06/2024 12:15:45          IMP   Provider          690

In [174]:
# Converting data types
merged_data['Offer Redemption Date'] = pd.to_datetime(merged_data['Offer Redemption Date'], errors='coerce')

# Calculate age based on 'Year of birth' and 'Offer Redemption Date'
current_year = datetime.datetime.now().year
merged_data['Age'] = current_year - merged_data['Year of birth']

# Removing duplicate entries
merged_data.drop_duplicates(inplace=True)

# Standardizing text data
merged_data['Call status'] = merged_data['Call status'].str.capitalize()
merged_data['Feedback'] = merged_data['Feedback'].str.capitalize()

merged_data.head()


Call status   Gender  Year of birth Visited facility Tiko card  \
0  Not reachable  Unknown         2006.0          Unknown   Unknown   
1  Not reachable  Unknown         2006.0          Unknown   Unknown   
2  Not reachable  Unknown         2006.0          Unknown   Unknown   
3  Not reachable  Unknown         2006.0          Unknown   Unknown   
4  Not reachable  Unknown         2006.0          Unknown   Unknown   

   Satisfaction      Feedback      Concerns Month       Offer Trader Name  \
0  Not provided  Not provided  Not provided  June      Nyalenda Gods Will   
1  Not provided  Not provided  Not provided  June      Nyalenda Gods Will   
2  Not provided  Not provided  Not provided  June      Nyalenda Gods Will   
3  Not provided  Not provided  Not provided  June  Doruex Medicare Centre   
4  Not provided  Not provided  Not provided  June  Doruex Medicare Centre   

  Enrolment Path Classifier            Enrolment Type  \
0            SMS - Low Tech  Enrolment With Mobiliser   
1            SMS - Low Tech  Enrolment With Mobiliser   
2            SMS - Low Tech  Enrolment With Mobiliser   
3            SMS - Low Tech  Enrolment With Mobiliser   
4            SMS - Low Tech  Enrolment With Mobiliser   

  Offer Trader County Name Offer Trader Business Organisation  \
0                   Kisumu                               Amua   
1                   Kisumu                               Amua   
2                   Kisumu                               Amua   
3                    Busia                                Rhn   
4                    Busia                                Rhn   

  Offer Redemption Date Service Type Actor Role  Tiko Amount   Age  
0   2024-06-26 12:14:28           OC      Agent          125  18.0  
1   2024-06-26 12:14:28           OC   Provider          150  18.0  
2   2024-06-26 12:14:28           OC     Rafiki           60  18.0  
3   2024-06-26 12:15:45          IMP      Agent           50  18.0  
4   2024-06-26 12:15:45          IMP   Provider          690  18.0

In [175]:
# Dropping the year of birth column to remain with the age column
del merged_data['Year of birth']

In [176]:
merged_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4075 entries, 0 to 4133
Data columns (total 18 columns):
 #   Column                              Non-Null Count  Dtype         
---  ------                              --------------  -----         
 0   Call status                         4075 non-null   object        
 1   Gender                              4075 non-null   object        
 2   Visited facility                    4075 non-null   object        
 3   Tiko card                           4075 non-null   object        
 4   Satisfaction                        4075 non-null   object        
 5   Feedback                            4075 non-null   object        
 6   Concerns                            4075 non-null   object        
 7   Month                               4075 non-null   object        
 8   Offer Trader Name                   4075 non-null   object        
 9   Enrolment Path Classifier           4075 non-null   object        
 10  Enrolment Type               

After cleaning, I have 4,075 entries (down from 4,134, likely due to duplicate removal).

In [177]:
# Exporting the cleaned dataframe
merged_data.to_excel("../data/clean_merged_data.xlsx", index=False)

Here's what I've done in a nutshell;

1. Cleaned the three datasets, call center data, tiko platform and financial impact data.
2. I proceeded to join the three datasets into one that I named merged_data.
3. I then went ahead to investigate the merged dataset and did more cleaning in the following steps;
     - Cleaned the 'Year of birth' column by replacing year ranges with approximate median values and converting it to numeric.
     - Filled missing values in various columns with appropriate placeholders.
     - Converted 'Offer Redemption Date' to datetime format.
     - Calculated the age of individuals based on their year of birth and the current year.
4. Lastly I exported the cleaned data and named it as clean_merged_data, that I will proceed to analyze in the visualization notebook.
